### ISSUE TITLE
(Swift) Crash trying to deallocate code connection property

### ISSUE THREAD

**Issue by cocojoe** [cocos2d/cocos2d-objc#1052]
Setup a code connection in SpriteBuilder, setting the `doc root var` and then do something like replace the scene.

```
import Foundation

class MainScene: CCNode {

    var _node: CCNode?

    func reload() {
        var scene: CCScene = CCBReader.loadAsScene("MainScene")
        CCDirector.sharedDirector().replaceScene(scene);
    }

}
```

**Comment by cocojoe** [comment]
Comment out the _node so not assigned with `object_setIvar` in `CCBReader.m` and no problem.

Error:
attempted to remove unregistered weak referrer 0x1009286c8 

Perhaps I'm just being a swift noob however feel like I've tried all the Swift side combos here to no avail.

Super simple example project:
https://www.dropbox.com/s/s92o9m6qugsvtyt/swiftpropertytest.spritebuilder.zip?dl=0

**Comment by slembcke** [comment]
Oof. I have no idea.

```
// 'target' is the Swift object, 'node' is the node set by SB.
(lldb) p target
(id) $0 = 0x00007f9348e60730
(lldb) p node
(CCNode *) $1 = 0x00007f9348e623e0

// The log when I press the 'Hello' button:
2014-11-05 04:36:43.193 swiftpropertytest[78136:1914897] Hello
2014-11-05 04:36:55.470 swiftpropertytest[78136:1914897] Hello
2014-11-05 04:36:55.470 swiftpropertytest[78136:1914897] cocos2d: deallocing <CCScene = 0x7f9348c676b0 | Name = (null)>
2014-11-05 04:36:55.470 swiftpropertytest[78136:1914897] cocos2d: deallocing <swiftpropertytest.MainScene = 0x7f9348e60730 | Name = >
2014-11-05 04:36:55.471 swiftpropertytest[78136:1914897] cocos2d: deallocing <CCNode = 0x7f9348e623e0 | Name = >

// The crash for me is happening in objc_msgSend()
// The object it's trying to message (presumably to call release) is the SB node.
// It was just dealloced above...
(lldb) p/x $rdi
(unsigned long) $1 = 0x00007f9348e623e0

// Up the stack a few frames is the dealloc method for an NSArray. (a CCNode.children array?)
// Up a few frams from that is the Swift node dealloc method.
(lldb) p self
(CCNode *) $2 = 0x00007f9348e60730
```

So something weird is happening with retaining counts. It seems that setting a Swift ivar with `object_setIvar()` does not cause it to retain. Haven't there already been a few working projects?

I'll try and make a more focused test later. I should not have stayed up this late. O_o

**Comment by slembcke** [comment]
Well... maybe not. It is calling `objc_storeStrong()` inside of `object_setIvar()`. Though this is all at the assembly level and I'm only half certain I'm reading it right. Bah.

**Comment by cocojoe** [comment]
I don't think many people have played around with SB and Swift given support only just became available and even then probably just single scene testing.  I've asked on Apple Forum.

**Comment by slembcke** [comment]
Made a bit of headway with this after poking around in the disassembler. `object_setIvar()` calls `class_getIvarLayout()` on the Swift class to determine if it should store a strong reference or not. This function seems to always return `NULL` for Swift objects. I feel like this is probably a bug, but the docs for what an "Ivar layout" is seems to be non-existant other than the runtime code itself (didn't bother to dig that much). If it is a bug and Apple fixes it we would be adding memory leaks by adding a manual retain call.

Another solution that applies to SB in Objective-C as well is to allow setting a property or an ivar. I would actually find this useful. In one case in Galactic Guardian I wished that the CCBReader would call property setters. Swift member variables always seem to synthesize Obj-C property setters. I feel like this would be a pretty hack-free solution that would fix other problems as well.

Making a patch now.

**Comment by slembcke** [comment]
Ugh. Not going so well. My first thought was to use KVC, but that doesn't have a way to check for a certain property or to fail gracefully if the property doesn't exist. Next up, generate a selector for the setter. Easy enough "name" becomes "setName:" right? Except the weird case that came up right away here is that "_node" becomes "set_node:". Even that isn't generally true since you can override the default name synthesis. If the property has a custom name, you can get that with some C code, but still have to generate it yourself if it doesn't. Finally, [NSString capitalizedString] capitalizes the first alpha character, and not the first character in the string.

Basically, time for a dinner break. Then write the overly complicated, fully correct version that does annoying amounts of C string fiddling.

**Comment by cocojoe** [comment]
Great work @slembcke nice bonus that can calls setters now.

**Comment by slembcke** [comment]
\o/ Woo!

(/me grumbles incoherently about the code being super gross)

**Comment by cocojoe** [comment]
Well at least Apple have now finally confirmed it was a bug.
